app/src/main/java/com/example/triptracker/viewmodel/HomeViewModel.kt

```diff
@@ -207,4 +203,71 @@ class HomeViewModel(private val repository: ItineraryRepository = ItineraryRepos
   fun setSearchFilter(filterType: FilterType) {
     _selectedFilter.value = filterType
   }
+
+  /**
+   * Filter the itinerary list by trending itineraries based on the flame count The trending
+   * itineraries are sorted in descending order of flame count
+   */
+  fun filterByTrending() {
+    _itineraryList.value =
+        _itineraryList.value?.sortedByDescending { itinerary -> itinerary.flameCount }
+  }
+
+  private fun calculateFlameCounts(saves: Long, clicks: Long, numStarts: Long): Long {
+    return 2 * saves + clicks + 5 * numStarts
+  }
+
+  /**
+   * Calculate the flame count for each itinerary and update the itineraryList the flame count is
+   * calculated as 2 * saves + clicks + 5 * numStarts as number of user that started the itinerary
+   * is presumably much smaller than the number of users that saved it and even less than the number
+   * of clicks and more valuable. updates the db field flameCount
+   *
+   * @param itineraryId the id of the itinerary to update
+   */
+  private fun updateFlameCount(itineraryId: String) {
+    repository.getItineraryById(itineraryId) { itinerary ->
+      if (itinerary != null) {
+        val newFlameCount =
+            calculateFlameCounts(itinerary.saves, itinerary.clicks, itinerary.numStarts)
+        repository.updateField(itineraryId, "flameCount", newFlameCount)
+      }
+    }
+  }
+
+  /** Increment the click count of the itinerary with the given id */
+  fun incrementClickCount(itineraryId: String) {
+    viewModelScope.launch {
+      repository.incrementField(itineraryId, "clicks")
+      // update flame count after incrementing click count
+      updateFlameCount(itineraryId)
+    }
+  }
+
+  /** Increment the save count of the itinerary with the given id */
+  fun incrementSaveCount(itineraryId: String) {
+    viewModelScope.launch {
+      repository.incrementField(itineraryId, "saves")
+      val updatedList =
+          _itineraryList.value?.map { itinerary ->
+            if (itinerary.id == itineraryId) {
+              itinerary.copy(saves = itinerary.saves + 1)
+            } else {
+              itinerary
+            }
+          }
+      if (updatedList != null) {
+        _itineraryList = MutableLiveData(updatedList)
+      }
+      updateFlameCount(itineraryId)
+    }
```

Comment:
Is the goal of theses functions the same as the one of the incrementField one ?

Reply:
yes but to update live Data, not the db !